### ISSUE TITLE
vesuvius.predict has no chunk cache: inference re-downloads the same chunks, ~17x more traffic than needed

### ISSUE THREAD

**Issue by aistae** [ScrollPrize/villa#1325]
## Summary

The inference read path reads every patch with an independent request to the
store. There is no chunk cache. Patches overlap (50% by default) and do not
start on chunk boundaries, so the same chunks are fetched again and again.

Measured on `main` (Python 3.14, installed from source), byte counter
installed on `zarr.storage.FsspecStore.get`, reading through the package's own
`VCDataset`.

## Reproduction

Reading **the same patch twice**:

```
patch #0, first read        16.78 MB,  8 requests
patch #0, read again        16.78 MB,  8 requests     <- nothing cached
adjacent patch (50% overlap) 16.78 MB,  8 requests
distant patch               16.78 MB,  8 requests
back to patch #0            16.78 MB,  8 requests
```

Volume: `PHercParis4/volumes/20260411134726-2.400um-0.2m-78keV-masked.zarr/2`,
`patch_size=(128,128,128)`, `step_size=0.5`, `skip_empty_patches=False`,
chunks 128³, `"compressor": null` so each chunk is exactly 2 097 152 bytes.

## Scale

| configuration | patches | chunk fetches | downloaded | unique chunks | minimum | waste |
|---|---|---|---|---|---|---|
| patch 128³, step 64 | 216 | 1728 | **3.62 GB** | 100 | **210 MB** | **17×** |
| patch 192³, step 96 (defaults) | 512 | 5832 | **12.2 GB** | 343 | **0.72 GB** | **17×** |

In the first case the useful payload of all patches is 0.45 GB, and 3.62 GB is
transferred to obtain it.

## Why this looks like a regression

Caching is used elsewhere in this repository:

- `neural_tracing/datasets/common.py` — fsspec `filecache` via
  `volume_cache_dir`, chunks cached across epochs
- `neural_tracing/infer.py` — `cachetools.LRUCache`

The inference path (`data/vc_dataset.py`, `data/utils.py::open_zarr`,
`models/run/inference.py`) has none — no `LRUStoreCache`, `LRUCache`,
`cache_type`, `filecache` or `blockcache`.

And `neural_tracing/heatmap_single_point/dataset.py` records how it was lost:

```python
if config.get("use_volume_store_cache", False):
    warnings.warn(
        "use_volume_store_cache is no longer supported with zarr 3 "
        "(LRUStoreCache was removed). Ignoring the option.",
        stacklevel=2,
    )
```

The cache relied on `zarr.LRUStoreCache`; zarr 3 removed that class, and the
option became a no-op.

## Suggested fix

The mechanism already exists here — apply the `neural_tracing` approach to the
main path:

1. fsspec `filecache` in `storage_options` for remote volumes (survives
   restarts), or
2. an LRU cache wrapping the zarr store, replacing the removed
   `LRUStoreCache`

## Secondary: traversal order is not chunk-aligned

Once a cache exists, the order in which patches are visited starts to matter.
Positions are enumerated row-major (`vc_dataset.py:335-338`) with
`DataLoader(shuffle=False)`, which does not match the chunk layout.

Simulated over a 384×8174×8174 region (patch 192³, step 96, 21 168 patches),
16 GB cache:

| traversal | traffic |
|---|---|
| row-major (current) | 60.1 GB |
| Morton / Z-order | **25.8 GB** |
| theoretical minimum | 25.77 GB |

Validated against a real transfer at smaller scale — simulated 201 MB vs
measured 201.3 MB for row-major, 120 vs 123.7 MB for Z-order (within 3%),
wall-clock −37%.

Checks: write locality also improves (54–57%); `num_workers` batching makes no
difference (identical for group sizes 1…64); but with `--num_parts >= 4` the
benefit disappears, because z-slab parts become thinner than one chunk
(`vc_dataset.py:382`) — so splitting into compact 3D blocks rather than
z-slabs would be the accompanying change.

I did not open a PR yet because the right shape of the fix is a design call —
happy to prepare one in whichever direction you prefer.

## Environment notes

Installing to reproduce this needed some detours, which may be worth separate
issues:

- `pip install vesuvius` on 3.14 resolves to `0.1.10` (Dec 2024) — that is
  already #1300
- the `volume-only` extra still pulls the local `volume-cartographer` path
  dependency, so the package cannot be installed standalone without the
  monorepo checkout
- `nest_asyncio`, `tqdm`, `scipy` are imported but not declared

Full write-up, scripts and raw numbers: https://github.com/aistae/vesuvius-io-audit

<sub>Investigation and write-up with Claude Code (Opus 5). Every figure above
is from an executed run against the live open-data bucket; the one simulated
result is labelled and validated against a measured transfer.</sub>

**Comment by TAUIL-Abd-Elilah** [comment]
Independent corroboration from the other end of the same problem — I measured this last week while checking whether whole-scroll streaming inference is affordable at all, and got the same effect at a larger scale.

Mine is a simulation over a real cross-section rather than live transfers: a 1536 × 7593 × 7593 slab of PHerc1218 at villa's inference defaults (192³, step 0.5) over 128³ uncompressed chunks — 93,615 patches touching 43,200 distinct chunks, **30.5× with no cache**. Your 17× at 128³/step-64 and this are the same effect; the ratio grows with patch size because a 192³ patch straddles more chunk boundaries.

One number that might be worth having before the design call gets made: at scroll scale the traversal order buys about as much as the cache size, and it is free.

| traversal | 4 GiB LRU | 16 GiB LRU |
|---|---|---|
| plain zyx | 2.83× | 1.97× |
| chunk-blocked | **2.17×** | **1.66×** |

Quadrupling the cache from 4 to 16 GiB buys less than reordering the patch list does. Same direction as your Morton measurement, and it suggests the ordering is not really a follow-up — a modest cache with the right order beats a large cache with the wrong one.

Why the difference matters rather than just being nice: my link caps at 11 MB/s (measured — 48 and 96 workers both return ~11, so it is the connection, not concurrency). PHerc1218 is 1.34 TB, so uncached that is 41 TB and about 43 days of I/O; at chunk-blocked with a 16 GiB cache it is 2.2 TB and about 56 h. That is the difference between an overnight job twice over and not possible at all from here.

Script and raw numbers are `bench_stream_inference.py` and `results/stream_read_amplification.json` in https://github.com/TAUIL-Abd-Elilah/vesuvius-repro. Happy to rerun it against whatever traversal you settle on.

**Comment by aistae** [comment]
@TAUIL-Abd-Elilah thank you — and you're right that I had the priority backwards.

**First, your numbers reproduce exactly.** I ran your configuration through my simulator and initially got 91,260 patches against your 93,615, so I went looking for my mistake. It was mine: I was generating positions with a fixed stride, while `compute_steps_for_sliding_window` takes `ceil((size - patch) / step) + 1` positions and distributes them *uniformly* so the last patch meets the edge. For 7593 that is 79 positions at an actual stride of 94.885, not 78 at 96. After fixing it:

| | mine | yours |
|---|---|---|
| patches | 93,615 | 93,615 |
| distinct chunks | 43,200 | 43,200 |
| fetches, no cache | 1,319,506 | 1,319,506 |
| amplification | 30.54× | 30.54× |

Your cache/order table too — 2.83×, 1.97×, 2.17×, 1.66× all land on the same values here. Worth noting the real stride being fractional makes alignment *worse* than my earlier model assumed, which is part of why my 17× understated it.

**On ordering being a follow-up: you've convinced me.** I had written it as secondary in this issue and said I'd rather land the cache first. Your point that quadrupling the cache buys less than reordering — and that reordering is free — is the right framing, and I've reordered the work to match: [#1327](https://github.com/ScrollPrize/villa/pull/1327) now carries both.

**One addition to your table.** A Morton curve over chunk indices does better than chunk-blocked, because it keeps locality across all three axes rather than resetting at each block boundary:

| traversal | 4 GiB | 16 GiB |
|---|---|---|
| zyx | 2.83× | 1.97× |
| chunk-blocked | 2.17× | 1.66× |
| **Morton** | **1.16×** | **1.05×** |

At 16 GiB that is 1.05×, essentially the floor of fetching each chunk once. Applied to your PHerc1218 arithmetic at 11 MB/s: 2.2 TB / ~56 h for chunk-blocked, against roughly 1.4 TB / ~35 h for Morton — though that is my extrapolation over your measured link, not something I can verify from here.

The implementation sorts `all_positions` before the split across parts and ranks, so each worker still receives a contiguous run of the curve. The patch set is unchanged, only the sequence, so outputs are identical — verified on `VCDataset`.

I'd welcome the rerun you offered. If it's easy on your side, the interesting comparison would be Morton against chunk-blocked at 4 GiB, where the gap should be widest.

<sub>Claude Code (Opus 5). The reproduction above was executed; the 35 h figure is arithmetic on your measured throughput.</sub>

**Comment by TAUIL-Abd-Elilah** [comment]
Ran it. **Morton reproduces here**, and it is a bigger win than chunk-blocked at every cache size:

| traversal | 1 GiB | 4 GiB | 16 GiB |
|---|---|---|---|
| zyx | 3.64× | 2.83× | 1.97× |
| chunk-blocked | 2.94× | 2.17× | 1.66× |
| **Morton** | **1.46×** | **1.19×** | **1.05×** |

At 16 GiB we agree exactly on 1.05×. **At 4 GiB I get 1.19× against your 1.16×** — small, but I would rather name it than round it away. My key is `morton(cz, cy, cx)` over the chunk index of the patch *start*, with the LRU counted per chunk touch; if you are keying on something slightly different (patch centre, or interleaving before the division) that would explain a gap of exactly this size. Not worth chasing unless it moves the decision, which it does not.

Two corrections to my own arithmetic while I am here. PHerc1218 is 23,247 × 7,593 × 7,593 = **1.34 TB**, not the 1.25 I quoted — an old hardcoded constant in my print, so every wall-clock I gave was 7% low. Corrected, at 11 MB/s:

| policy | fetched | I/O |
|---|---|---|
| no cache | 40.9 TB | 1,034 h |
| chunk-blocked, 4 GiB | 2.90 TB | 73 h |
| chunk-blocked, 16 GiB | 2.22 TB | 56 h |
| **Morton, 4 GiB** | **1.59 TB** | **40 h** |
| **Morton, 16 GiB** | **1.41 TB** | **36 h** |

So your ~35 h lands, and the more useful number is that **Morton at 4 GiB beats chunk-blocked at 16 GiB** — the ordering is worth more than 4× the memory.

**And a first end-to-end datapoint on your "does it help in a real run" question**, though it raises a question rather than answering it. I ran villa's own predict against the remote store over a 384³ region of PHerc1218 at level 0, no local copy, TTA off, batch 1: **216 patches in 239 s, 0.90 patches/s, peak 24,018 MiB of a 24,576 MiB card.** The striking part is that the region spans only 216 chunks, so even at full amplification this is far under the 11 MB/s the link can sustain — it is not bandwidth-bound, and 1.1 s per patch is far slower than the forward pass. That points at per-request latency rather than bytes, which if it holds up makes your *request-count* reduction worth more than the byte reduction, and makes Morton better still.

One run, so I am stating it as an observation and not a finding. The cache-on/cache-off pair over the identical region is next and I will post it here either way.

Script, Morton implementation and raw numbers: https://github.com/TAUIL-Abd-Elilah/vesuvius-repro at `6e28e3e` (`bench_stream_inference.py`, `bench_gpu_stream_gate.py`).

**Comment by TAUIL-Abd-Elilah** [comment]
Posting this either way, as promised — and the honest answer is that **my instrument is not good enough to answer the question yet, and the observation in my previous comment does not survive its own control.**

I ran the identical configuration four more times: same 384³ region of PHerc1218, level 0, TTA off, batch 1, 216 patches, nothing else changed.

| arm | wall-clock | patches/s | peak GPU |
|---|---|---|---|
| no cache (run 1) | 239.3 s | 0.90 | 24,018 MiB |
| **no cache (run 2)** | **41.5 s** | **5.20** | 23,167 MiB |
| simplecache, cold | 45.9 s | 4.71 | 22,912 MiB |
| **no cache (run 3)** | **253.7 s** | **0.85** | 23,919 MiB |

**Three runs of one configuration span 41.5 s to 253.7 s — about 6×, which is far larger than any cache effect I was trying to measure.** So:

**I retract both inferences from my last comment.** "Not bandwidth-bound" and "this points at per-request latency" both rested on the 239 s run being representative, and it is not. The 41.5 s run moves 453 MB of spanned chunks in 41.5 s — **10.9 MB/s, essentially this machine's measured link ceiling** — which is exactly what a *fully* bandwidth-bound run looks like. The two readings are opposite and I picked one from a single sample.

**The cache arm means nothing here.** 45.9 s sits inside the no-cache spread. I am also dropping the warm-cache arm rather than reporting it: it came back with a 2,410 MiB GPU peak against ~23,000 MiB for every other arm, so I do not believe it did the same work, and a number I can't explain is not a number.

**One thing that is worth saying, because it bounds where my own simulation applies.** A 384³ region touches 216 chunks — 453 MB — which fits in RAM, so the OS holds it and there is nothing for a chunk cache to recover. Read amplification only bites once the traversal's working set exceeds memory, and one z-slab of PHerc1218 is 7.0 GiB. **A small ROI test structurally cannot show the effect the simulation is about**, which is my error in experiment design, not a disagreement with your numbers.

What I'll fix before posting anything else on this: interleave the arms instead of running them in blocks, n ≥ 5 per arm, a concurrent bandwidth probe recorded alongside each run so a slow link is visible rather than inferred, and a region large enough that the working set actually exceeds RAM.

**None of this touches the amplification table** — that is an exact count over the chunk grid, not a timing, and it stands as posted.

**Comment by TAUIL-Abd-Elilah** [comment]
Took my own advice from this thread and dropped the clock. Read amplification is a count, not a duration, so instead of timing runs I wrapped `HTTPFileSystem._cat_file` and counted the chunk requests the real inference path actually issues.

**The simulation is exact.** Same 256³ bbox of PHerc1218, level 0, run for real against the bucket:

| | fetches | distinct chunks | amplification |
|---|---|---|---|
| simulated | 1,728 | 125 | **13.824×** |
| measured | 1,728 | 125 | **13.824×** |
| measured, replicate | 1,728 | 125 | **13.824×** |

Identical to the individual fetch, twice. So the 30.5× whole-slab figure earlier in this thread is a model that reproduces real HTTP traffic rather than approximating it — and the smaller number here is just region size, which the simulator predicts too.

**The method also has no run-to-run variance**, which is the actual point. Those same runs on the clock were the ones that gave me 41.5 s, 239 s and 254 s. Failed fetches raise and never get counted, so a link falling over costs wall-clock and changes the count by nothing.

**I still cannot measure the cache arm, and the failure mode is worth flagging.** The local simplecache shim I was using dies under zarr 3 with `Timeout context manager should be used inside a task`, leaving an empty cache directory and a failed run — so its zero-fetch reading is the instrument breaking, not the cache working. Which is exactly what a working cache would look like on this counter, so anyone measuring one this way should check the run actually completed. #1177's in-store LRU is the right thing to count; if it lands I will run it through the same counter and post the number.

One gotcha for anyone copying this: with `--num_workers > 0` the reads happen in DataLoader worker processes, the parent-process patch does not exist there, and the run cheerfully reports zero fetches. Counting needs `--num_workers 0`.

`bench_io_counted.py` and `run_predict_counted.py` in https://github.com/TAUIL-Abd-Elilah/vesuvius-repro
